sopel/modules/url.py

```diff
@@ -297,14 +297,20 @@ def title_auto(bot: SopelWrapper, trigger: Trigger):
     if re.match(bot.config.core.prefix + r'\S+', trigger):
         return
 
-    # Avoid fetching known malicious links
-    if 'safety_cache' in bot.memory and trigger in bot.memory['safety_cache']:
-        if bot.memory['safety_cache'][trigger]['positives'] > 1:
-            return
-
-    urls = web.search_urls(
+    unchecked_urls = web.search_urls(
         trigger, exclusion_char=bot.config.url.exclusion_char, clean=True)
 
+    urls = []
+    for url in unchecked_urls:
+        # Avoid fetching known malicious links
+        if url in bot.memory.get("safety_cache", {}):
+            if bot.memory["safety_cache"][url]["positives"] > 0:
+                continue
+        hostname = urlparse(url).hostname.lower()
+        if hostname in bot.memory.get("safety_cache_local", {}):
+            continue
```

Comment:
Each `dict.get` or `dict[...]` call cost a little more than doing this:

```suggestion
    urls = []
    safety_cache = bot.memory.get("safety_cache", {})
    safety_cache_local = bot.memory.get("safety_cache_local", {})
    for url in unchecked_urls:
        # Avoid fetching known malicious links
        if url in safety_cache and safety_cache[url]["positives"] > 0:
            continue
        hostname = urlparse(url).hostname.lower()
        if hostname in safety_cache_local:
            continue
```

And yes, it could be a list-comprehension. Or put that in a function that would `yield url`... so many options. A for-loop is fine.